core/src/main/java/com/linecorp/armeria/internal/RequestContextUtil.java

```diff
@@ -87,5 +87,20 @@ public static SafeCloseable pushWithRootAndOldCtx(ClientRequestContext currentCt
         }
     }
 
+    /**
+     * An {@link IllegalStateException} which is raised when pushing a context from the unexpected thread
+     * or forgetting to close the previous context.
+     */
+    public static final class IllegalContextPushingException extends IllegalStateException {
+
+        private static final long serialVersionUID = 5431942355463120798L;
+
+        public IllegalContextPushingException(RequestContext newCtx, RequestContext oldCtx) {
+            super("Trying to call object wrapped with context " + newCtx + ", but context is currently " +
+                  "set to " + oldCtx + ". This means the callback was called from " +
+                  "unexpected thread or forgetting to close previous context.");
```

Comment:
Now that we have a dedicated exception type, we could add getters for the two contexts? We could also simplify the exception message and add more details to the class Javadoc.